[JIT] adding torch.jit.isinstance support

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46062 [JIT] adding torch.jit.isinstance support**

Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

Differential Revision: [D24264415](https://our.internmc.facebook.com/intern/diff/D24264415)

Commit:
[WIP] adding torch.jit.isinstance

[ghstack-poisoned]

Commit:
Update on "[WIP] adding torch.jit.isinstance"


Currently eager implementation only works for python 3.8+ 

Also have issue of shadowing isinstance within the eager implementation, which is why it is torch.jit.isinstance2 

[ghstack-poisoned]

Commit:
Update on "[WIP] adding torch.jit.isinstance"


have issue of shadowing isinstance within the eager implementation, which is why it is torch.jit.isinstance2 

[ghstack-poisoned]

Commit:
Update on "[WIP] adding torch.jit.isinstance"



[ghstack-poisoned]

Lilyjjo commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -1196,6 +1196,14 @@ struct to_ir {
+              checkApplyNumInputs(apply, 2);
+              return emitIsInstance(apply.inputs()[0], apply.inputs()[1]);
+            }
+          }
```
This is where the type refinement is happening. I'm not sure if what I did to extract the identify of the torch.jit.isinstance case with lines 1199-1202 is the cleanest way that it could be done

gmagogsfm commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,109 @@
+        )
+
+
+def generics_checker(the_obj, the_type):
```
Doesn't have to be in this diff.

Could you add a TODO for implementing check for "future"?

gmagogsfm commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,109 @@
+    origin_type = get_origin(the_type)
+    if origin_type:
+        return generics_checker(the_obj, the_type)
+    # handle odd case of non typed optional origin returning as none
```
OOC, why is this the case? Is this a quirk of Python?

Lilyjjo replied:
It's an oddity of how python treats the Optional type. For the plain type Optional without an inner specification (e.g. Optional[str]), calling `getattr(the_type, "__origin__", None)` returns None. This is opposed to Dict/List/Tuple with no inner specification where the returned origin is a list/dict/tuple instead of None

gmagogsfm commented on test/jit/test_isinstance.py:
```diff
@@ -0,0 +1,293 @@
+
+        x = ["1", "2", "3"]
+
+        try:
```
You can use `self.assertRaisesRegex` or `self.assertRaisesRegexWithHighlight` to check for exception instead of writing `try-catch`

Lilyjjo replied:
the self.assertRaisesRegex (from what I can tell) only works with runtime errors generated in the C++ code. Since these errors are generated in Python I wasn't able to get them to work with the self.assertRaisesRegex- do you know if this assumption of mine is wrong?

Lilyjjo replied:
I really are having issues getting self.assertRaisesRegex to work. 

`with self.assertRaisesRegex(
                RuntimeError, 
                "Attempted to use Tuple without a "
                "contained type. Please add a contained type, e.g. "
                "Tuple[int]"
        ):
            torch.jit.script(tuple_no_contained_type`

I'm adding this to the end of the tuple test down below on line 240 and it fails to work. I've tried formatting the error return message string in all the ways I can think of. I also can't find a use of self.assertRaisesRegex which deals with Python-generated errors, the errors they handle in other use cases are all generated from c++ from what I can tell

gmagogsfm commented on test/jit/test_isinstance.py:
```diff
@@ -0,0 +1,293 @@
+                for val in obj.values():
+                    str_cat = str_cat + val
+                assert "111222" == str_cat
+            assert hit
```
Thanks! I didn't see it during my first quick (30 seconds) scan that day.

Lilyjjo commented on torch/jit/__init__.py:
```diff
@@ -70,5 +74,9 @@ def annotate(the_type, the_value):
     return the_value
 
 
+def isinstance(the_obj, the_type) -> bool:
```
I was thinking that torch.jit.isinstance(...) was a better name than torch.jit._isinstance(...), but I do think I can change it to _isinstance if you think that would be the better name

gmagogsfm replied:
Oh, I meant you don't have to implement the name change via adding a wrapper call. Instead you can literally write 
```
isinstance = _isinstance
```

here to save one function call.

Lilyjjo commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,109 @@
+    return getattr(the_type, "__args__", None)
+
+
+def check_args_exist(the_type):
```
yes, can do

Lilyjjo replied:
So actually, looking more into this, the is_list/is_dict/etc in _jit_internal.py do more checks that what is necessary for the actual torch.jit.isinstance needs. The first check in the _jit_internal.py which triggers the shared error message is the only real things that _isintance needs, the return values from the _jit_internal.py aren't needed. I see the benefit of using the _jit_internal.py functions because the error messages should be the same and it is code duplication to have them twice. I can see a couple options:

1. use the _jit_internal.py functions in _isinstance and perform the unnecessary checks 
2. re-write the error messages in a new function which both the _jit_intenral.py functions and  _isinstnace functions can utilize
3. keep the code as is in this PR

Lilyjjo commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,109 @@
+                    return False
+        else:
+            return False
+    elif origin_type is Union:  # TODO actually handles Optional Case
```
Calling `getattr(the_type, "__origin__", None)` on an Optional[str] type returns the type of typing.Union instead of typing.Option. Whoever wrote the is_optional logic in _jit_internal.py did some similarly odd logic to address this

Commit:
Update on "[JIT] adding torch.jit.isinstance support"


Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

[ghstack-poisoned]

SplitInfinity commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,107 @@
+            "contained types. Please add contained type, e.g. "
+            "Dict[int, int]"
+        )
+    elif target_type is None or target_type is Optional:
```
The inclusion of `target_type is None` is bit confusing. `torch.jit.isinstance(x, None)` seems like perfectly valid code to me. It seems to be a bit too tightly coupled to `check_args_exist` and the weird behaviour you noted below where an `Optional` with no contained type has no `__origin__`.

gmagogsfm replied:
`None` is actually not a type in Python and `isinstance(x, None)` is not allowed. Or is there a specific use case you are thinking about?

SplitInfinity replied:
Oh that's true, usually `x is None` is used for that.

SplitInfinity commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,107 @@
+def check_args_exist(target_type):
+    if target_type is List or target_type is list:
+        raise RuntimeError(
+            "Attempted to use List without a "
+            "contained type. Please add a contained type, e.g. "
+            "List[int]"
```
Consider creating a template string for this and reusing it. All of these error messages look very similar except for type names.

Lilyjjo replied:
Is there a strong reason to create a template string? I don't think this version of the explicit string is bad

SplitInfinity commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,107 @@
+    elif origin_type is list or origin_type is List:
+        if not isinstance(obj, list):
+            return False
+        for el in obj:
+            # check if nested generics, ex: List[List[str]]
+            arg_type = get_args(target_type)[0]
+            arg_origin = get_origin(arg_type)
```
```suggestion
        arg_type = get_args(target_type)[0]
        arg_origin = get_origin(arg_type)
        for el in obj:
            # check if nested generics, ex: List[List[str]]

```

SplitInfinity commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,107 @@
+    check_args_exist(target_type)
+    if origin_type is None:
+        pass
+    elif origin_type is list or origin_type is List:
```
Out of curiosity, under what circumstances can `origin_type` be `typing.List`?

Lilyjjo replied:
In python 3.6 it's typing.List

SplitInfinity commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,107 @@
+def generics_checker(obj, target_type):
+    origin_type = get_origin(target_type)
+    check_args_exist(target_type)
+    if origin_type is None:
+        pass
```
Which case does this handle? All callsites of `generics_checker` I see have nearby calls to `get_origin` and `generics_checker` is called only if `get_origin` retursn something.

Lilyjjo replied:
My b, yeah that was left over from a different implementation

SplitInfinity commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,107 @@
+            return True
+        else:
+            return False
+    return True
```
What is the base case for?

Lilyjjo replied:
I was trying to use the logic that if the code passed the checks then it must be true, but then I didn't consider that different containers could be passed through besides List/Dict/Tuple/Optional. I fixed the logic, thanks

SplitInfinity commented on test/jit/test_isinstance.py:
```diff
@@ -0,0 +1,293 @@
+        except RuntimeError as e:
+            self.assertEqual(
+                str(e),
+                "Attempted to use List without a "
+                "contained type. Please add a contained type, e.g. "
+                "List[int]",
```
Consider using a common template string for all of these error checks; the messages look quite similar except for the types involved.

Lilyjjo commented on torch/jit/__init__.py:
```diff
@@ -1,4 +1,8 @@
 import torch._C
+from torch.jit._isinstance import _isinstance
+from typing import List, Dict, Optional, Tuple, Union
+import typing
+from sys import version_info
```
My b again, I originally had all the logic in this file and forgot to delete this part

Lilyjjo commented on torch/jit/_isinstance.py:
```diff
@@ -0,0 +1,107 @@
+from typing import List, Dict, Tuple, Union, Optional
+
+
+def get_origin(target_type):
+    return getattr(target_type, "__origin__", None)
+
+
+def get_args(target_type):
+    return getattr(target_type, "__args__", None)
```
I think the code works here because in 3.6 `List` is treated differently from `List[str]` : 

```
Python 3.6.2 |Anaconda, Inc.| (default, Oct  5 2017, 07:59:26) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import List
>>> a = List[str]
>>> a.__args__
(<class 'str'>,)
>>> a.__origin__
typing.List
>>> a = List
>>> a.__args__
>>> a.__origin__
```

But this allows eager mode to accept List/Dict/Tuple as the object type while Scripting doesn't for python 3.6. I'll see what I can do to make the behavior the same

Commit:
Update on "[JIT] adding torch.jit.isinstance support"


Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

Differential Revision: [D24264415](https://our.internmc.facebook.com/intern/diff/D24264415)

[ghstack-poisoned]

Lilyjjo commented on torch/jit/_script.py:
```diff
@@ -1079,5 +1080,100 @@ def _unwrap_optional(x):
+    elif target_type is Dict or target_type is dict:
+        _jit_internal.is_dict(target_type)
+    elif target_type is None or target_type is Optional:
+        _jit_internal.is_optional(target_type)
```
Here I'm using the _jit_internal.py functions only for the error message they throw, the return value of these function calls aren't used. I'm still questioning if I should use the _jit_internal.py functions this way or if I should extract the error messages from the _jit_internal.py functions, put them in a different location, and then have this part of the code for _isinstance and the _jit_internal.py functions call them separately

gmagogsfm replied:
I think either way is fine with a slight preference to refactor the error message part and reuse. Just to make sure I understand why we don't use those return values, it is because you are rolling your own logic to check `origin` and `typing` like on line 1108 and 1109?

Lilyjjo replied:
The _jit_internal.py functions don't aid in helping to check for nested containers, they only check if the outer container is of the right type and if the outer container contains an inner type. In generics_checker on 1102-1157 there's a need to do work to process nested containers and I think the code is cleaner when I don't use the _jit_intneral.py functions.

I'm also confused on what the checks for the `__module__` below code int he _jit_internal.py functions are doing:
```
if not hasattr(ann, '__module__'):
        return False
    return ann.__module__ == 'typing' and \
        (getattr(ann, '__origin__', None) is List or
            getattr(ann, '__origin__', None) is list)
```
Should I be checking for this `__module__` attribute in my logic somewhere? I'm not familiar enough yet with modules to know if I'm missing something here

gmagogsfm replied:
I don't think this is necessary since you are already checking equality against `List`, which is from `typing` module. I am not entirely sure why it is needed in _jit_internal.py. @SplitInfinity Could you clarify?

SplitInfinity replied:
Yeah, I'm not sure either.

I think it might also make sense to move this code to `jit_internal.py` to reuse those error messages. I would prefer that to calling `is_list`, etc. for the error message and throwing away the return value. `ignore`, `export`, `is_scripting` and `ignore` live there and those are certainly not internal APIs (but I think they are in that file in order to avoid circular import problems).

Lilyjjo replied:
Okay, I'll move the code and factor out the error messages 💯

Lilyjjo commented on torch/jit/_script.py:
```diff
@@ -1079,5 +1080,100 @@ def _unwrap_optional(x):
+
+    # handle python 3.6 quality of returning None for origin of 
+    # containers without contained type (vs returning container type)
+    if version_info[1] == 6:
```
```suggestion
    check_args_exist(target_type) 
```
Here I could just call `check_args_exist(target_type)` no matter what the version number and the code behaves the same

Commit:
Update on "[JIT] adding torch.jit.isinstance support"


Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

Differential Revision: [D24264415](https://our.internmc.facebook.com/intern/diff/D24264415)

[ghstack-poisoned]

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -1079,5 +1079,99 @@ def _unwrap_optional(x):
+    if origin_type:
+        return generics_checker(obj, target_type)
+
+    # Check to handle weird python type behaviors
+    # 1. python 3.6 returns None for origin of containers without 
+    #    contained type (intead of returning outer container type)
+    # 2. non-typed optional origin returns as none instead 
+    #    of as optional in 3.6-3.8
```
Where is the check? Is it between 1148 and 1156?

Lilyjjo replied:
The check happens in check_args_exist on line 1170, the two above cases fall through the lines on 1162-1163 even though they are containers that are missing the contained type

gmagogsfm commented on test/jit/test_isinstance.py:
```diff
@@ -0,0 +1,246 @@
+        def optional_test_none(x: Any):
+            assert torch.jit.isinstance(x, Optional[torch.Tensor])
+            # assert not torch.jit.isinstance(x, Optional[str])
+            # TODO: above line fails in TS interpreter need to investigate
```
Are you looking into this?

Lilyjjo replied:
I'm running into the difficulty where eager mode and scripting behave differently. Eager mode will allow both isinstances to return true (since the None can be either in Optional[torch.Tensor] or Optional[str] Python), but from what I can tell during scripting type refinement happens during the first isinstance call refining the type to Optional[torch.Tensor] which then makes the 2nd isinstance call return false since the type isn't Optional[str]. 

I'm not sure how to go about resolving this. I don't know how to in python eager mode carry the type assignment outcome between isinstance calls to make the 2nd isinstance call return false

gmagogsfm replied:
Good finding.

I don't think we should change eager mode behavior as we would want users to mechanically migrate from `isinstance` to `torch.jit.isinstance`.

I think what we can do though is to change behavior of `prim::isinstance` in TorchScript. From what I can tell, `prim::isinstance` ultimate calls into [here](https://github.com/pytorch/pytorch/blob/38c97fb6f0725fa0c02687023b101f0f8339cca7/torch/csrc/jit/runtime/vararg_functions.cpp#L106) which checks for sub-typing relationship. I think this logic is only correct when all types either completely overlaps (subtype relationship) or do not overlap (not an instance of). `Optional[T]` doesn't fall into this situation. It would be even worse for `Union[T1, T2, T3]` when we support it in the future.

We should therefore fix this by changing implementation of `prim::isinstance`. It doesn't have to be in this diff though since this case is rare and unconventional enough. Feel free to leave a TODO here for now and we can discuss how to deal with it with more audience.

Lilyjjo commented on test/jit/test_isinstance.py:
```diff
@@ -0,0 +1,246 @@
+        def optional_test(x: Any):
+            assert torch.jit.isinstance(x, Optional[torch.Tensor])
+            assert not torch.jit.isinstance(x, Optional[str])
+            # TODO: successful torch.jit.isinstance makes sets type?
```
I'll delete this TODO 😬

gmagogsfm reviewed (approved):
Looks good, approving. But please wait for SplitIfinity to clarify on one of the comment before merging.

SplitInfinity reviewed (commented):
Could you add docstrings for the new functions that you introduced?

Commit:
Update on "[JIT] adding torch.jit.isinstance support"


Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

Differential Revision: [D24264415](https://our.internmc.facebook.com/intern/diff/D24264415)

[ghstack-poisoned]

Lilyjjo commented on torch/_jit_internal.py:
```diff
@@ -885,3 +869,135 @@ def _is_exception(obj):
+                        print(val)
+
+        m = torch.jit.script(MyModule())
+    """
```
@SplitInfinity is this what you had in mind for a doc string? I'm not exactly clear on what doc strings should look like.

Another doc string question: should this go here or on should it be added to the torch/jit/init.py callsite on line 75 of that file?

SplitInfinity replied:
I think this particular docstring with example code that uses `torch.jit.isinstance` should go on `isinstance` in `torch/jit/__init__.py`.

It's also a good idea to include sections describing the arguments and return values. For example, here is the docstring from `torch.jit.wait`:
```
def wait(future):
    """
    Forces completion of a `torch.jit.Future[T]` asynchronous task, returning the
    result of the task. See :func:`~fork` for docs and examples.
    Arguments:
        func (torch.jit.Future[T]): an asynchronous task reference, created through `torch.jit.fork`
    Returns:
        `T`: the return value of the the completed task
    """
```

SplitInfinity commented:
Also, can you take a look at the coverage report and add some tests for a few lines that aren't being covered if possible? For example, I think the line that returns `False` when a value in a dictionary isn't of the expected type isn't being hit: https://codecov.io/gh/pytorch/pytorch/pull/46062/diff#D1-1132

Commit:
Update on "[JIT] adding torch.jit.isinstance support"


Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

Differential Revision: [D24264415](https://our.internmc.facebook.com/intern/diff/D24264415)

[ghstack-poisoned]

Commit:
Update on "[JIT] adding torch.jit.isinstance support"


Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

Differential Revision: [D24264415](https://our.internmc.facebook.com/intern/diff/D24264415)

[ghstack-poisoned]

Commit:
Update on "[JIT] adding torch.jit.isinstance support"


Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

Differential Revision: [D24264415](https://our.internmc.facebook.com/intern/diff/D24264415)

[ghstack-poisoned]

Commit:
Update on "[JIT] adding torch.jit.isinstance support"


Adds support for torch.jit.isinstance in both eager and script mode

Example use:

```
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.isinstance(input, List[str]):
            for el in input:
                print(el)

TestModule().forward(["1","2"])
scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```

Differential Revision: [D24264415](https://our.internmc.facebook.com/intern/diff/D24264415)

[ghstack-poisoned]
